src/dotty/tools/dotc/core/TyperState.scala

```diff
@@ -59,6 +59,20 @@ class TyperState(r: Reporter) extends DotClass with Showable {
   /** Commit state so that it gets propagated to enclosing context */
   def commit()(implicit ctx: Context): Unit = unsupported("commit")
 
+  /** The typer state has already been committed */
+  def isCommitted: Boolean = false
+
+  /** Optionally, if this is a mutable typerstate, it's creator state */
+  def parent: Option[TyperState] = None
+
+  /** The closest ancestor of this typer state (including possible this typer state itself)
+   *  which is not yet committed.
+   */
+  def uncommittedAncestor: TyperState = parent match {
+    case Some(p) if p.isCommitted => p.uncommittedAncestor
+    case _ => this
```

Comment:
If `this` is commited but its parent is not committed, we end up in this branch and return `this` instead of its parent, shouldn't we do something like this instead:
```scala
def uncommittedAncestor: Option[TyperState] =
  if (!this.isCommitted)
    Some(this)
  else
    parent.flatMap(_.isCommitted)
```